src/compression/mod.rs

```diff
@@ -0,0 +1,394 @@
+            #[cfg(feature = "snappy-compression")]
+            Compressor::Snappy => write!(f, "snappy"),
+        }
+    }
+}
+
+impl<'de> Deserialize<'de> for Compressor {
+    fn deserialize<D: Deserializer<'de>>(deserializer: D) -> std::result::Result<Self, D::Error> {
+        let s = String::deserialize(deserializer)?;
+        s.parse().map_err(|e| D::Error::custom(format!("{}", e)))
+    }
+}
+
+impl Serialize for Compressor {
+    fn serialize<S>(&self, serializer: S) -> std::result::Result<S::Ok, S::Error>
+    where
+        S: Serializer,
+    {
+        self.to_string().serialize(serializer)
+    }
+}
+
+#[cfg(any(
+    feature = "zstd-compression",
+    feature = "zlib-compression",
+    feature = "snappy-compression"
+))]
+impl Compressor {
+    pub(crate) fn id(&self) -> CompressorId {
+        match *self {
+            #[cfg(feature = "zstd-compression")]
+            Compressor::Zstd { level: _ } => CompressorId::Zstd,
+            #[cfg(feature = "zlib-compression")]
+            Compressor::Zlib { level: _ } => CompressorId::Zlib,
+            #[cfg(feature = "snappy-compression")]
+            Compressor::Snappy => CompressorId::Snappy,
+        }
+    }
+
+    pub(crate) fn validate(&self) -> Result<()> {
+        match *self {
+            #[cfg(feature = "zstd-compression")]
+            Compressor::Zstd { level: Some(level) }
+                if !zstd::compression_level_range().contains(&level) =>
+            {
+                Err(Error::from(ErrorKind::InvalidArgument {
+                    message: format!("invalid zstd level: {}", level),
+                }))
+            }
+            #[cfg(feature = "zlib-compression")]
+            Compressor::Zlib { level: Some(level) } if !(-1..10).contains(&level) => {
+                Err(Error::from(ErrorKind::InvalidArgument {
+                    message: format!("invalid zlib level: {}", level),
+                }))
+            }
+            _ => Ok(()),
+        }
+    }
+
+    pub(crate) fn to_encoder(&self) -> Result<Encoder> {
+        match *self {
+            #[cfg(feature = "zstd-compression")]
+            Compressor::Zstd { level } => {
+                let encoder =
+                    zstd::Encoder::new(vec![], level.unwrap_or(zstd::DEFAULT_COMPRESSION_LEVEL))
+                        .map_err(|e| {
+                            Error::from(ErrorKind::Internal {
+                                message: format!(
+                                    "an error occurred getting a new zstd encoder: {}",
+                                    e
+                                ),
+                            })
+                        })?;
+
+                Ok(Encoder::Zstd { encoder })
+            }
+            #[cfg(feature = "zlib-compression")]
+            Compressor::Zlib { level } => {
+                let level = match level {
+                    Some(level) if level != -1 => Compression::new(level as u32),
```

Comment:
If a user provides a negative value that isn't `-1`, this cast will likely produce an incorrect value. Instead, can we use `TryFrom` here and map the error to our `InvalidArgument` error?

Reply:
We can do that, although any invalid values would have been caught by now.